MdeModulePkg/Core/Dxe/DxeMain/DxeMain.c

```diff
@@ -778,7 +780,10 @@ CoreExitBootServices (
   // Notify other drivers of their last chance to use boot services
   // before the memory map is terminated.
   //
-  CoreNotifySignalList (&gEfiEventBeforeExitBootServicesGuid);
+  if (!mExitBootServicesCalled) {

```

Comment:
We should also only call the EBS handlers once per UEFI spec, right:

UEFI spec 2.10 errata A section 7.4.6
“All events from the EFI_EVENT_GROUP_BEFORE_EXIT_BOOT_SERVICES and EFI_EVENT_GROUP_EXIT_BOOT_SERVICES event notification groups as well as events of type EVT_SIGNAL_EXIT_BOOT_SERVICES must be signaled before ExitBootServices() returns EFI_SUCCESS. The events are only signaled once even if ExitBootServices() is called multiple times.“

Reply:
Ah yes, thanks for digging that up. So for EFI_EVENT_GROUP_BEFORE_EXIT_BOOT_SERVICES, it is literally in the spec.

I think this is implied for EFI_EVENT_GROUP_EXIT_BOOT_SERVICES, though: once `CoreExitBootServices ()` decides to proceed, i.e., when the map key is correct, it will run to completion and clear the `BootServices` pointer in the system table, and so calling ExitBootServices() again is unlikely to happen. So I don't think this is something to worry about.

Reply:
Ah yes, I see that now.